In [18]:
import pandas as pd

In [19]:
df = pd.read_csv('../data/US-Airline-Flight-Routes-and-Fares-1993-2024.csv')

df.head()

/var/folders/2s/6s11c01n27g08gwkl3rs_2dc0000gn/T/ipykernel_21776/3419103000.py:1: DtypeWarning: Columns (20,21) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/US-Airline-Flight-Routes-and-Fares-1993-2024.csv')


tbl  Year  quarter  citymarketid_1  citymarketid_2  \
0  Table1a  2021        3           30135           33195   
1  Table1a  2021        3           30135           33195   
2  Table1a  2021        3           30140           30194   
3  Table1a  2021        3           30140           30194   
4  Table1a  2021        3           30140           30466   

                            city1                          city2  airportid_1  \
0  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)        10135   
1  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)        10135   
2                 Albuquerque, NM          Dallas/Fort Worth, TX        10140   
3                 Albuquerque, NM          Dallas/Fort Worth, TX        10140   
4                 Albuquerque, NM                    Phoenix, AZ        10140   

   airportid_2 airport_1  ...    fare  carrier_lg  large_ms  fare_lg  \
0        14112       ABE  ...   81.43          G4    1.0000    81.43   
1        15304       ABE  ...  208.93          DL    0.4659   219.98   
2        11259       ABQ  ...  184.56          WN    0.9968   184.44   
3        11298       ABQ  ...  182.64          AA    0.9774   183.09   
4        14107       ABQ  ...  177.11          WN    0.6061   184.49   

  carrier_low   lf_ms  fare_low Geocoded_City1  Geocoded_City2  \
0          G4  1.0000     81.43            NaN             NaN   
1          UA  0.1193    154.11            NaN             NaN   
2          WN  0.9968    184.44            NaN             NaN   
3          AA  0.9774    183.09            NaN             NaN   
4          AA  0.3939    165.77            NaN             NaN   

                 tbl1apk  
0  202131013514112ABEPIE  
1  202131013515304ABETPA  
2  202131014011259ABQDAL  
3  202131014011298ABQDFW  
4  202131014014107ABQPHX  

[5 rows x 23 columns]

Running .info() on our data frame to get insight on our data

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 245955 entries, 0 to 245954
Data columns (total 23 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   tbl             245955 non-null  object 
 1   Year            245955 non-null  int64  
 2   quarter         245955 non-null  int64  
 3   citymarketid_1  245955 non-null  int64  
 4   citymarketid_2  245955 non-null  int64  
 5   city1           245955 non-null  object 
 6   city2           245955 non-null  object 
 7   airportid_1     245955 non-null  int64  
 8   airportid_2     245955 non-null  int64  
 9   airport_1       245955 non-null  object 
 10  airport_2       245955 non-null  object 
 11  nsmiles         245955 non-null  int64  
 12  passengers      245955 non-null  int64  
 13  fare            245955 non-null  float64
 14  carrier_lg      244415 non-null  object 
 15  large_ms        244415 non-null  float64
 16  fare_lg         244415 non-null  float64
 17  carrier_lo

Checking our current columns to ensure naming follows a specific structure. Also using this to determine if we could have better column names making our data easier to work with.

In [21]:
df.columns

Index(['tbl', 'Year', 'quarter', 'citymarketid_1', 'citymarketid_2', 'city1',
       'city2', 'airportid_1', 'airportid_2', 'airport_1', 'airport_2',
       'nsmiles', 'passengers', 'fare', 'carrier_lg', 'large_ms', 'fare_lg',
       'carrier_low', 'lf_ms', 'fare_low', 'Geocoded_City1', 'Geocoded_City2',
       'tbl1apk'],
      dtype='object')

Dropping columns we do not need, as well as renaming the majority of columns to be more fitting and easier to work with.

In [22]:
df.drop({'citymarketid_1', 'citymarketid_2', 'Geocoded_City1', 'Geocoded_City2', 'Geocoded_City1', 'Geocoded_City2'}, axis=1, inplace=True)
df.rename(columns={'quarter' : 'Quarter', 'city1' : 'Origin_City', 'city2' : 'Dest_City', 
                   'airportid_1' : 'Orgin_AiportId', 'airportid_2' : 'Dest_AirportId', 
                   'airport_1' : 'Origin_Airport_Code', 'airport_2' : 'Dest_Airport_Code',
                   'nsmiles' : 'Trip_Miles', 'passengers' : "Passengers", 'fare' : "Avg_Fare", 'carrier_lg' : "Largest_Carrier", 'large_ms' : 'Largest_Carrier_MS',
                   'fare_lg' : 'Avg_Fare_Largest_Carrier', 'carrier_low' : 'Lowest_Fare_Carrier_Code', 'fare_low' : 'Lowest_Fare', 'lf_ms' : 'Lowest_Fare_Carrier_MS'
                   }, inplace=True)
df.sample(10    )

tbl  Year  Quarter                            Origin_City  \
208406  Table 1a  2018        2         Boston, MA (Metropolitan Area)   
201408  Table 1a  2017        2    Los Angeles, CA (Metropolitan Area)   
121339  Table 1a  2006        3                              Boise, ID   
169019  Table 1a  2008        2  New York City, NY (Metropolitan Area)   
206908  Table 1a  2018        1  New York City, NY (Metropolitan Area)   
95366   Table 1a  1999        2        Atlanta, GA (Metropolitan Area)   
193455  Table 1a  2016        3                      Oklahoma City, OK   
48548   Table 1a  2011        4         Boston, MA (Metropolitan Area)   
229864   Table1a  2021        2                       Indianapolis, IN   
73821   Table 1a  2003        1                         Fort Myers, FL   

                                    Dest_City  Orgin_AiportId  Dest_AirportId  \
208406               Minneapolis/St. Paul, MN           10721           13487   
201408        Norfolk, VA (Metropolitan Area)           13891           14098   
121339  San Francisco, CA (Metropolitan Area)           10713           14771   
169019     Washington, DC (Metropolitan Area)           12391           12264   
206908  San Francisco, CA (Metropolitan Area)           11618           14771   
95366       Cleveland, OH (Metropolitan Area)           10397           10874   
193455  San Francisco, CA (Metropolitan Area)           13851           14771   
48548                             Seattle, WA           13296           14747   
229864    Los Angeles, CA (Metropolitan Area)           12339           14908   
73821   New York City, NY (Metropolitan Area)           14635           12478   

       Origin_Airport_Code Dest_Airport_Code  Trip_Miles  Passengers  \
208406                 BOS               MSP        1124        1705   
201408                 ONT               PHF        2371           1   
121339                 BOI               SFO         523         169   
169019                 ISP               IAD         265           1   
206908                 EWR               SFO        2619        3161   
95366                  ATL               CAK         554         331   
193455                 OKC               SFO        1384         123   
48548                  MHT               SEA        2496          53   
229864                 IND               SNA        1814          93   
73821                  RSW               JFK        1120         925   

        Avg_Fare Largest_Carrier  Largest_Carrier_MS  \
208406    172.44              DL              0.5800   
201408    569.29              AA              0.8571   
121339    192.61              UA              0.9300   
169019    445.98              US              0.8000   
206908    407.81              UA              0.7626   
95366     113.31              FL              0.8100   
193455    299.46              UA              0.5727   
48548     276.35              WN              0.3500   
229864    249.37              AA              0.4605   
73821     138.51              B6              0.9400   

        Avg_Fare_Largest_Carrier Lowest_Fare_Carrier_Code  \
208406                    203.01                       SY   
201408                    599.33                       DL   
121339                    192.38                       AS   
169019                    477.97                       US   
206908                    449.07                       VX   
95366                     109.43                       FL   
193455                    310.78                       WN   
48548                     277.17                       DL   
229864                    250.98                       WN   
73821                     139.29                       DL   

        Lowest_Fare_Carrier_MS  Lowest_Fare                tbl1apk  
208406                  0.1887       118.51  201821072113487BOSMSP  
201408                  0.1429       389.00  201721389114098ONTPHF  
121339           

## Handling missing data

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 245955 entries, 0 to 245954
Data columns (total 19 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   tbl                       245955 non-null  object 
 1   Year                      245955 non-null  int64  
 2   Quarter                   245955 non-null  int64  
 3   Origin_City               245955 non-null  object 
 4   Dest_City                 245955 non-null  object 
 5   Orgin_AiportId            245955 non-null  int64  
 6   Dest_AirportId            245955 non-null  int64  
 7   Origin_Airport_Code       245955 non-null  object 
 8   Dest_Airport_Code         245955 non-null  object 
 9   Trip_Miles                245955 non-null  int64  
 10  Passengers                245955 non-null  int64  
 11  Avg_Fare                  245955 non-null  float64
 12  Largest_Carrier           244415 non-null  object 
 13  Largest_Carrier_MS        244415 non-null  f

We have 245,955 rows of data. Our total rows with at least one missing value is ~1,612. Thats only ~0.65% of our data. In this case we can just drop these rows with minimal impact to our data. 

In [ ]:
for col in df.select_dtypes(include='object').columns:
    print(f"{col}: {df[col].unique()[:10]}")
